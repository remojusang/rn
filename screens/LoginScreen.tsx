import { useRef, useEffect, useState } from 'react';
import type { NativeStackScreenProps } from '@react-navigation/native-stack';
import { View, StyleSheet, TextInput, Alert } from 'react-native';
import FormInput from '../components/FormInput';
import { RootStackParamList } from '../utils/types';
import { useForm } from 'react-hook-form';
import CustomBtn from '../components/CustomBtn';
import FormLayout from '../components/FormLayout';
import FlexGap from '../components/FlexGap';
import { isLoggedInState, uidState } from '../components/Atoms';
import { useSetRecoilState } from 'recoil';
import { signIn } from '../utils/firebase/auth';
import { AUTH_MSG } from '../utils/constants';
import ErrorAlert from '../components/ErrorAlert';

export type LoginProps = NativeStackScreenProps<
  RootStackParamList,
  'Login'
>;

interface ILoginForm {
  email: string;
  password: string;
}

function LoginScreen({ route, navigation }: LoginProps) {
  const { formState, handleSubmit, control, setValue } =
    useForm<ILoginForm>({
      mode: 'onChange',
    });
  useEffect(() => {
    if (route.params?.signedEmail) {
      setValue('email', route.params.signedEmail);
    }
  }, [navigation, route, setValue]);
  const setisLoggedIn = useSetRecoilState(isLoggedInState);
  const setUid = useSetRecoilState(uidState);
  const pwRef = useRef<TextInput>(null);
  const [signInLoading, setSignInLoading] = useState(false);
  const onValid = async ({ email, password }: ILoginForm) => {
    try {
      setSignInLoading(true);
      const {
        user: { uid },
      } = await signIn({ email, password });
      setUid(uid);
      setisLoggedIn(true);
      Alert.alert('Notification', '로그인 완료.', [{ text: 'OK' }]);
    } catch (e: any) {
      const err_msg = AUTH_MSG[e.code] || 'login failed';
      ErrorAlert(err_msg);
    } finally {
      setSignInLoading(false);
    }
  };
  return (
    <FormLayout>
      <>
        <FormInput
          name="email"
          errorMsg={formState.errors.email?.message}
          control={control}
          textInputConfig={{
            keyboardType: 'email-address',
            placeholder: 'remo@naver.com',
          }}
          onNext={() => pwRef.current?.focus()}
        />
        <FormInput
          name="password"
          errorMsg={formState.errors.password?.message}
          control={control}
          textInputConfig={{
            placeholder: '********',
            secureTextEntry: true,
          }}
          inputRef={pwRef}
        />
        <FlexGap gapSize={10} />
        <CustomBtn
          isLoading={signInLoading}
          title="loginBtn"
          onPress={handleSubmit(onValid)}
        />
        <FlexGap gapSize={10} />
        <CustomBtn
          isLoading={false}
          title="googleBtn"
          isGoogle
          onPress={handleSubmit(onValid)}
        />
        <View style={styles.divider} />
        <CustomBtn
          isLoading={false}
          title="signUpBtn"
          color="#6C6C6C"
          onPress={() => navigation.navigate('Sign')}
        />
      </>
    </FormLayout>
  );
}

export default LoginScreen;

const styles = StyleSheet.create({
  gap: {
    marginVertical: 5,
  },
  divider: {
    width: '100%',
    height: 1,
    backgroundColor: 'black',
    marginVertical: 10,
  },
});
