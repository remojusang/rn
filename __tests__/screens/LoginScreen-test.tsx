import { render } from '@testing-library/react-native';
import LoginScreen from '../../screens/LoginScreen';

describe('LoginScreen', () => {
  test('렌더링 테스트', () => {
    const TEST_PROPS: any = {
      route: jest.fn(),
      navigation: jest.fn(),
    };
    const renderedJson = render(
      <LoginScreen {...TEST_PROPS} />,
    ).toJSON();
    expect(renderedJson).toMatchSnapshot();
    expect(renderedJson).toBeTruthy();
  });
});
