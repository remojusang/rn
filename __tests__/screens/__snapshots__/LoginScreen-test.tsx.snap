// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginScreen 렌더링 테스트 1`] = `
<RCTSafeAreaView
  style={
    Object {
      "alignItems": "center",
      "flex": 1,
      "justifyContent": "center",
      "margin": 20,
    }
  }
>
  <View
    collapsable={false}
    style={
      Object {
        "marginVertical": 16,
        "transform": Array [
          Object {
            "translateX": 0,
          },
        ],
        "width": "100%",
      }
    }
  >
    <View
      style={
        Object {
          "flexDirection": "row",
          "justifyContent": "space-between",
          "marginBottom": 10,
        }
      }
    >
      <Text
        style={
          Object {
            "color": "black",
            "fontSize": 16,
            "fontWeight": "bold",
          }
        }
      >
        이메일
      </Text>
      <Text
        style={
          Object {
            "fontWeight": "300",
          }
        }
      />
    </View>
    <TextInput
      accessibilityHint=""
      autoCapitalize="none"
      keyboardType="email-address"
      onChangeText={[Function]}
      placeholder="remo@naver.com"
      style={
        Object {
          "backgroundColor": "white",
          "borderColor": "#000",
          "borderRadius": 6,
          "borderWidth": 1,
          "fontSize": 18,
          "padding": 10,
        }
      }
    />
  </View>
  <View
    collapsable={false}
    style={
      Object {
        "marginVertical": 16,
        "transform": Array [
          Object {
            "translateX": 0,
          },
        ],
        "width": "100%",
      }
    }
  >
    <View
      style={
        Object {
          "flexDirection": "row",
          "justifyContent": "space-between",
          "marginBottom": 10,
        }
      }
    >
      <Text
        style={
          Object {
            "color": "black",
            "fontSize": 16,
            "fontWeight": "bold",
          }
        }
      >
        비밀번호
      </Text>
      <Text
        style={
          Object {
            "fontWeight": "300",
          }
        }
      />
    </View>
    <TextInput
      accessibilityHint=""
      autoCapitalize="none"
      onChangeText={[Function]}
      placeholder="********"
      secureTextEntry={true}
      style={
        Object {
          "backgroundColor": "white",
          "borderColor": "#000",
          "borderRadius": 6,
          "borderWidth": 1,
          "fontSize": 18,
          "padding": 10,
        }
      }
    />
  </View>
  <View
    style={
      Object {
        "marginVertical": 5,
      }
    }
  />
  <View
    style={
      Object {
        "width": "100%",
      }
    }
  >
    <View
      accessible={true}
      collapsable={false}
      focusable={true}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "backgroundColor": "#2F2F33",
          "borderRadius": 10,
          "padding": 12,
        }
      }
    >
      <Text
        style={
          Object {
            "color": "white",
            "fontSize": 16,
            "fontWeight": "bold",
            "textAlign": "center",
          }
        }
      >
        로그인
      </Text>
    </View>
  </View>
  <View
    style={
      Object {
        "marginVertical": 5,
      }
    }
  />
  <View
    style={
      Object {
        "width": "100%",
      }
    }
  >
    <View
      accessible={true}
      collapsable={false}
      focusable={true}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "backgroundColor": "#4649FF",
          "borderRadius": 10,
          "padding": 12,
        }
      }
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "flexDirection": "row",
            "justifyContent": "center",
          }
        }
      >
        <Image
          source={
            Object {
              "testUri": "../../../assets/google_logo.png",
            }
          }
          style={
            Object {
              "marginRight": 10,
            }
          }
        />
        <Text
          style={
            Object {
              "color": "white",
              "fontSize": 16,
              "fontWeight": "bold",
              "textAlign": "center",
            }
          }
        >
          Google 계정으로 로그인
        </Text>
      </View>
    </View>
  </View>
  <View
    style={
      Object {
        "backgroundColor": "black",
        "height": 1,
        "marginVertical": 10,
        "width": "100%",
      }
    }
  />
  <View
    style={
      Object {
        "width": "100%",
      }
    }
  >
    <View
      accessible={true}
      collapsable={false}
      focusable={true}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "backgroundColor": "#6C6C6C",
          "borderRadius": 10,
          "padding": 12,
        }
      }
    >
      <Text
        style={
          Object {
            "color": "white",
            "fontSize": 16,
            "fontWeight": "bold",
            "textAlign": "center",
          }
        }
      >
        회원가입
      </Text>
    </View>
  </View>
</RCTSafeAreaView>
`;
